src/optimise/train.jl

```diff
@@ -120,16 +131,15 @@ The callback can call [`Flux.stop`](@ref) to interrupt the training loop.
 
 Multiple callbacks can be passed to `cb` as array.
 """
-function train!(loss, ps::Params, data, opt::AbstractOptimiser; cb = () -> ())
+function train!(loss, ad::AD.AbstractBackend, model, data, optstate; cb = () -> ())
   cb = runall(cb)
   itrsz = Base.IteratorSize(typeof(data))
   n = (itrsz == Base.HasLength()) || (itrsz == Base.HasShape{1}()) ? length(data) : 0
   @withprogress for (i, d) in enumerate(data)
     try
-      gs = gradient(ps) do
-        loss(batchmemaybe(d)...)
-      end
-      update!(opt, ps, gs)
+      _loss = _build_loss(ad, loss, batchmemaybe(d))
+      gs = _gradient_only(AD.gradient(ad, _loss, model))
+      optstate, model = update(optstate, model, gs)
```

Comment:
```suggestion
      optstate, model = update!(optstate, model, gs)
```
I think we can get away with this by default.